<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ rCASC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rCASC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">5.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rCASC_vignette.html">rCASC vignette</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="anovaLike.html">anovaLike()</a></code> </p>
        </td>
        <td><p>A function allowing the identification of differentially expressed genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="autoFeature.html">autoFeature()</a></code> </p>
        </td>
        <td><p>Feature selection for autoencoder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="autoencoder.html">autoencoder()</a></code> </p>
        </td>
        <td><p>Autoencoder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="autoencoderAnalysis.html">autoencoderAnalysis()</a></code> </p>
        </td>
        <td><p>Autoencoder Analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="autoencoderClustering.html">autoencoderClustering()</a></code> </p>
        </td>
        <td><p>Autoencoder clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bootstrapsVideo.html">bootstrapsVideo()</a></code> </p>
        </td>
        <td><p>Bootstraps vide</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bulkClusters.html">bulkClusters()</a></code> </p>
        </td>
        <td><p>bulkClusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ccRemove.html">ccRemove()</a></code> </p>
        </td>
        <td><p>Cc remove</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cellCycle2.html">cellCycle2()</a></code> </p>
        </td>
        <td><p>Cell Cycle</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cellrangerCount.html">cellrangerCount()</a></code> </p>
        </td>
        <td><p>Cellranger count</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cellrangerIndexing.html">cellrangerIndexing()</a></code> </p>
        </td>
        <td><p>Cellranger indexing</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkCountDepth.html">checkCountDepth()</a></code> </p>
        </td>
        <td><p>Running SCnorm  checkCountDepth test.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clusterIdentification.html">clusterIdentification()</a></code> </p>
        </td>
        <td><p>Cluster identification</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clusterNgriph.html">clusterNgriph()</a></code> </p>
        </td>
        <td><p>Defining with griph the range of  number of clusters to be used with SIMLR</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clusterReorg.html">clusterReorg()</a></code> </p>
        </td>
        <td><p>Reorganize Cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clusterStability.html">clusterStability()</a></code> </p>
        </td>
        <td><p>Permutations and Clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustersFeatures.html">clustersFeatures()</a></code> </p>
        </td>
        <td><p>A function to extract priority lists of genes from ANOVAlike output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cometsc.html">cometsc()</a></code> </p>
        </td>
        <td><p>Marker genes discovery with COMETSC</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cometsc2.html">cometsc2()</a></code> </p>
        </td>
        <td><p>Marker genes discovery with COMETSC for autoencoder results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="counts2log.html">counts2log()</a></code> </p>
        </td>
        <td><p>A function to convert raw count in log counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="crossLabel.html">crossLabel()</a></code> </p>
        </td>
        <td><p>Cross Label</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="csvToSparse.html">csvToSparse()</a></code> </p>
        </td>
        <td><p>csv to Sparse</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deDetection.html">deDetection()</a></code> </p>
        </td>
        <td><p>A function allowing the identification of differentially expressed genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deTwoGroups.html">deTwoGroups()</a></code> </p>
        </td>
        <td><p>A function allowing the identification of differentially expressed genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deltaFilter.html">deltaFilter()</a></code> </p>
        </td>
        <td><p>Filterin</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="demultiplexing.html">demultiplexing()</a></code> </p>
        </td>
        <td><p>Generating running bcl2fastq</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dimensions.html">dimensions()</a></code> </p>
        </td>
        <td><p>Cells count table size</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dockerTest.html">dockerTest()</a></code> </p>
        </td>
        <td><p>Testing if Docker is installed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downloadContainers.html">downloadContainers()</a></code> </p>
        </td>
        <td><p>Download for the first time all containers embedded in the workflows</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fastqc.html">fastqc()</a></code> </p>
        </td>
        <td><p>A function to handle a docker containier executing fastqc</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterZeros.html">filterZeros()</a></code> </p>
        </td>
        <td><p>A function plotting the distribution of zeros in cells eliminating all genes without a certain fraction of 0 in the samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genesPrioritization.html">genesPrioritization()</a></code> </p>
        </td>
        <td><p>Gene prioritization with SIMLR</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genesSelection.html">genesSelection()</a></code> </p>
        </td>
        <td><p>Genes selection from genesPrioritization output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genesUmi.html">genesUmi()</a></code> </p>
        </td>
        <td><p>Plotting genes to UMIs relationship.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="griphBootstrap.html">griphBootstrap()</a></code> </p>
        </td>
        <td><p>Executing clustering with griph.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="h5tocsv.html">h5tocsv()</a></code> </p>
        </td>
        <td><p>sparse to dense</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="harmony.html">harmony()</a></code> </p>
        </td>
        <td><p>Harmony</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hfc.html">hfc()</a></code> </p>
        </td>
        <td><p>hfc, heatmap for clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="houseKeep.html">houseKeep()</a></code> </p>
        </td>
        <td><p>houseKeep Filtering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="impute.html">impute()</a></code> </p>
        </td>
        <td><p>Imputing dropouts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="indropCounts.html">indropCounts()</a></code> </p>
        </td>
        <td><p>A function to handle a indrop V2 single cell data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="indropIndex.html">indropIndex()</a></code> </p>
        </td>
        <td><p>A function to create a genome index for indrop V2 single cell data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lorenzFilter.html">lorenzFilter()</a></code> </p>
        </td>
        <td><p>A function to handle sigle cell Lorenz Quality filter for Single-cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mergeMatrix.html">mergeMatrix()</a></code> </p>
        </td>
        <td><p>Merge matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mitoRiboUmi.html">mitoRiboUmi()</a></code> </p>
        </td>
        <td><p>MitoRiboUmi</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="nClusterEvaluationSIMLR.html">nClusterEvaluationSIMLR()</a></code> </p>
        </td>
        <td><p>Evaluation of the numbur of optimal clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permAnalysis.html">permAnalysis()</a></code> </p>
        </td>
        <td><p>Permutation Analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permAnalysisGriph.html">permAnalysisGriph()</a></code> </p>
        </td>
        <td><p>Permutation Analysis Griph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permAnalysisSeurat.html">permAnalysisSeurat()</a></code> </p>
        </td>
        <td><p>Permutation Analysis Seurat</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permutationClustering.html">permutationClustering()</a></code> </p>
        </td>
        <td><p>Permutations and Clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permutationClusteringSHARP.html">permutationClusteringSHARP()</a></code> </p>
        </td>
        <td><p>Permutations and Clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permutationMovie.html">permutationMovie()</a></code> </p>
        </td>
        <td><p>Permutation Movie</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recatPrediction.html">recatPrediction()</a></code> </p>
        </td>
        <td><p>Cell Cycle</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDocker.html">runDocker()</a></code> </p>
        </td>
        <td><p>Run docker container</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scannobyGtf.html">scannobyGtf()</a></code> </p>
        </td>
        <td><p>Annotating single cell counts table using ENSEMBL gtf and refGenome CRAN package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scanpyBootstrap.html">scanpyBootstrap()</a></code> </p>
        </td>
        <td><p>Executing clustering with scanpy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scanpyPermutation.html">scanpyPermutation()</a></code> </p>
        </td>
        <td><p>Scanpy Permutation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scnorm.html">scnorm()</a></code> </p>
        </td>
        <td><p>Running SCnorm  normalization</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratBootstrap.html">seuratBootstrap()</a></code> </p>
        </td>
        <td><p>Executing clustering with Seurat</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratPCAEval.html">seuratPCAEval()</a></code> </p>
        </td>
        <td><p>Estimating the range of PC components to be used in Seurat clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratPermutation.html">seuratPermutation()</a></code> </p>
        </td>
        <td><p>Seurat Permutation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratPrior.html">seuratPrior()</a></code> </p>
        </td>
        <td><p>seuratprior</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sharpBootstrap.html">sharpBootstrap()</a></code> </p>
        </td>
        <td><p>Executing clustering with SHARP.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simlrBootstrap.html">simlrBootstrap()</a></code> </p>
        </td>
        <td><p>Clustering with SIMLR</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spatialAnalysis.html">spatialAnalysis()</a></code> </p>
        </td>
        <td><p>spatial Analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spatialAnalysis2.html">spatialAnalysis2()</a></code> </p>
        </td>
        <td><p>spatial Analysis Plus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="splitClusters.html">splitClusters()</a></code> </p>
        </td>
        <td><p>Split Matrix for each clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sraDownload.html">sraDownload()</a></code> </p>
        </td>
        <td><p>A function to handle fasterq-dumper SRA to download SRA fastq files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stpipeline.html">stpipeline()</a></code> </p>
        </td>
        <td><p>Spatial transcriptomics pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subSetCell.html">subSetCell()</a></code> </p>
        </td>
        <td><p>Subset Cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="topx.html">topx()</a></code> </p>
        </td>
        <td><p>A function to selectec top X on the basis of gene/transcript expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tsneBootstrap.html">tsneBootstrap()</a></code> </p>
        </td>
        <td><p>Executing clustering with tSne</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="umiNorm.html">umiNorm()</a></code> </p>
        </td>
        <td><p>Running SCnorm  normalization</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unstableFiltering.html">unstableFiltering()</a></code> </p>
        </td>
        <td><p>unstableFiltering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vioHTMLByGenes.html">vioHTMLByGenes()</a></code> </p>
        </td>
        <td><p>vioHTMLByGenes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wrapperAutoencoder.html">wrapperAutoencoder()</a></code> </p>
        </td>
        <td><p>wrapperAutoencoder</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Luca Alessandri, Marco Beccuti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


