<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Annotating single cell counts table using ENSEMBL gtf and refGenome CRAN package — scannobyGtf • rCASC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Annotating single cell counts table using ENSEMBL gtf and refGenome CRAN package — scannobyGtf" />
<meta property="og:description" content="This function executes the docker container annotate.1, where refGenome is used to annotate a single cell counts table with ensembl gene ids on first column using ENSEMBL GTF annotation" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rCASC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rCASC_vignette.html">rCASC vignette</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Annotating single cell counts table using ENSEMBL gtf and refGenome CRAN package</h1>
    
    <div class="hidden name"><code>scannobyGtf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function executes the docker container annotate.1, where refGenome is used to annotate a single cell counts table with ensembl gene ids on first column using ENSEMBL GTF annotation</p>
    </div>

    <pre class="usage"><span class='fu'>scannobyGtf</span>(
  <span class='kw'>group</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"docker"</span>, <span class='st'>"sudo"</span>),
  <span class='no'>file</span>,
  <span class='no'>gtf.name</span>,
  <span class='kw'>biotype</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>mt</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>TRUE</span>, <span class='fl'>FALSE</span>),
  <span class='kw'>ribo.proteins</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>TRUE</span>, <span class='fl'>FALSE</span>),
  <span class='kw'>umiXgene</span> <span class='kw'>=</span> <span class='fl'>3</span>,
  <span class='kw'>riboStart.percentage</span> <span class='kw'>=</span> <span class='fl'>20</span>,
  <span class='kw'>riboEnd.percentage</span> <span class='kw'>=</span> <span class='fl'>70</span>,
  <span class='kw'>mitoStart.percentage</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>mitoEnd.percentage</span> <span class='kw'>=</span> <span class='fl'>100</span>,
  <span class='kw'>thresholdGenes</span> <span class='kw'>=</span> <span class='fl'>250</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>group, </th>
      <td><p>a character string. Two options: <code>"sudo"</code> or <code>"docker"</code>, depending to which group the user belongs</p></td>
    </tr>
    <tr>
      <th>file, </th>
      <td><p>a character string indicating the folder where input data are located and where output will be written and matrix name "/bin/users/matrix.csv". The system recognize automatically csv as comma separated files and txt as tab separated file</p></td>
    </tr>
    <tr>
      <th>gtf.name, </th>
      <td><p>a character string indicating the ENSEMBL gtf file</p></td>
    </tr>
    <tr>
      <th>biotype, </th>
      <td><p>a character string the biotypes of interest</p></td>
    </tr>
    <tr>
      <th>mt, </th>
      <td><p>a boolean to define if mitocondrial genes have to be removed, FALSE mean that mt genes are removed</p></td>
    </tr>
    <tr>
      <th>ribo.proteins, </th>
      <td><p>a boolean to define if ribosomal proteins have to be removed, FALSE mean that ribosomal proteins (gene names starting with rpl or rps) are removed</p></td>
    </tr>
    <tr>
      <th>umiXgene, </th>
      <td><p>a integer defining how many UMI are required to call a gene as present. default: 3</p></td>
    </tr>
    <tr>
      <th>riboStart.percentage, </th>
      <td><p>start range for ribosomal percentage, cells within the range are kept</p></td>
    </tr>
    <tr>
      <th>riboEnd.percentage, </th>
      <td><p>end range for ribosomal percentagem cells within the range are kept</p></td>
    </tr>
    <tr>
      <th>mitoStart.percentage, </th>
      <td><p>start range for mitochondrial percentage, cells within the range are retained</p></td>
    </tr>
    <tr>
      <th>mitoEnd.percentage, </th>
      <td><p>end range for mitochondrial percentage, cells within the range are retained</p></td>
    </tr>
    <tr>
      <th>thresholdGenes, </th>
      <td><p>parameter to filter cells according to the number og significative genes expressed</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>one file: annotated_counts table, where ensembl ids are linked to gene symbols and a PDF showing the effect of ribo and mito genes removal. Filtered_annotated annotated counts table with only cells and genes given by filtering thresholds. A pdf showing the effect of genes counts of the filtering and a filteredStatistics.txt indicating how many cell and genes were filtered out</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
        <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"wget http://130.192.119.59/public/testSCumi_mm10.csv.zip"</span>)
     <span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>rCASC</span>)
     <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"unzip testSCumi_mm10.csv.zip"</span>)
     <span class='co'>#filtering low quality cells</span>
     <span class='fu'><a href='lorenzFilter.html'>lorenzFilter</a></span>(<span class='kw'>group</span><span class='kw'>=</span><span class='st'>"docker"</span>,<span class='kw'>scratch.folder</span><span class='kw'>=</span><span class='st'>"/data/scratch/"</span>,
                  <span class='kw'>file</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span>(),<span class='st'>"testSCumi_mm10.csv"</span>,<span class='kw'>sep</span><span class='kw'>=</span><span class='st'>"/"</span>),
                  <span class='kw'>p_value</span><span class='kw'>=</span><span class='fl'>0.05</span>,<span class='kw'>separator</span><span class='kw'>=</span><span class='st'>','</span>)
     <span class='co'>#running annotation and removal of mit and ribo proteins genes</span>
     <span class='co'>#download mouse GTF for mm10</span>
     <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"wget ftp://ftp.ensembl.org/pub/release-92/gtf/mus_musculus/Mus_musculus.GRCm38.92.gtf.gz"</span>)
     <span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span>(<span class='st'>"gunzip Mus_musculus.GRCm38.92.gtf.gz"</span>)
     <span class='fu'>scannobyGtf</span>(<span class='kw'>group</span><span class='kw'>=</span><span class='st'>"docker"</span>, <span class='kw'>file</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span>(),<span class='st'>"testSCumi_mm10.csv"</span>,<span class='kw'>sep</span><span class='kw'>=</span><span class='st'>"/"</span>),
                  <span class='kw'>gtf.name</span><span class='kw'>=</span><span class='st'>"Mus_musculus.GRCm38.94.gtf"</span>, <span class='kw'>biotype</span><span class='kw'>=</span><span class='st'>"protein_coding"</span>,
                  <span class='kw'>mt</span><span class='kw'>=</span><span class='fl'>TRUE</span>, <span class='kw'>ribo.proteins</span><span class='kw'>=</span><span class='fl'>TRUE</span>, <span class='kw'>umiXgene</span><span class='kw'>=</span><span class='fl'>3</span>, <span class='kw'>riboStart.percentage</span><span class='kw'>=</span><span class='fl'>0</span>,
                  <span class='kw'>riboEnd.percentage</span><span class='kw'>=</span><span class='fl'>100</span>, <span class='kw'>mitoStart.percentage</span><span class='kw'>=</span><span class='fl'>0</span>, <span class='kw'>mitoEnd.percentage</span><span class='kw'>=</span><span class='fl'>100</span>, <span class='kw'>thresholdGenes</span><span class='kw'>=</span><span class='fl'>100</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Raffaele Calogero, Luca Alessandri</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Luca Alessandri, Marco Beccuti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


