<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Annotating single cell counts table using ENSEMBL gtf and refGenome CRAN package — scannobyGtf • rCASC</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Annotating single cell counts table using ENSEMBL gtf and refGenome CRAN package — scannobyGtf"><meta property="og:description" content="This function executes the docker container annotate.1, where refGenome is used to annotate a single cell counts table with ensembl gene ids on first column using ENSEMBL GTF annotation"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rCASC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/rCASC_vignette.html">rCASC vignette</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Annotating single cell counts table using ENSEMBL gtf and refGenome CRAN package</h1>
    
    <div class="hidden name"><code>scannobyGtf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function executes the docker container annotate.1, where refGenome is used to annotate a single cell counts table with ensembl gene ids on first column using ENSEMBL GTF annotation</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scannobyGtf</span><span class="op">(</span></span>
<span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"docker"</span>, <span class="st">"sudo"</span><span class="op">)</span>,</span>
<span>  <span class="va">file</span>,</span>
<span>  <span class="va">gtf.name</span>,</span>
<span>  biotype <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>  ribo.proteins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>  umiXgene <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  riboStart.percentage <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  riboEnd.percentage <span class="op">=</span> <span class="fl">70</span>,</span>
<span>  mitoStart.percentage <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mitoEnd.percentage <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  thresholdGenes <span class="op">=</span> <span class="fl">250</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>group, </dt>
<dd><p>a character string. Two options: <code>"sudo"</code> or <code>"docker"</code>, depending to which group the user belongs</p></dd>


<dt>file, </dt>
<dd><p>a character string indicating the folder where input data are located and where output will be written and matrix name "/bin/users/matrix.csv". The system recognize automatically csv as comma separated files and txt as tab separated file</p></dd>


<dt>gtf.name, </dt>
<dd><p>a character string indicating the ENSEMBL gtf file</p></dd>


<dt>biotype, </dt>
<dd><p>a character string the biotypes of interest</p></dd>


<dt>mt, </dt>
<dd><p>a boolean to define if mitocondrial genes have to be removed, FALSE mean that mt genes are removed</p></dd>


<dt>ribo.proteins, </dt>
<dd><p>a boolean to define if ribosomal proteins have to be removed, FALSE mean that ribosomal proteins (gene names starting with rpl or rps) are removed</p></dd>


<dt>umiXgene, </dt>
<dd><p>a integer defining how many UMI are required to call a gene as present. default: 3</p></dd>


<dt>riboStart.percentage, </dt>
<dd><p>start range for ribosomal percentage, cells within the range are kept</p></dd>


<dt>riboEnd.percentage, </dt>
<dd><p>end range for ribosomal percentagem cells within the range are kept</p></dd>


<dt>mitoStart.percentage, </dt>
<dd><p>start range for mitochondrial percentage, cells within the range are retained</p></dd>


<dt>mitoEnd.percentage, </dt>
<dd><p>end range for mitochondrial percentage, cells within the range are retained</p></dd>


<dt>thresholdGenes, </dt>
<dd><p>parameter to filter cells according to the number og significative genes expressed</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>one file: annotated_counts table, where ensembl ids are linked to gene symbols and a PDF showing the effect of ribo and mito genes removal. Filtered_annotated annotated counts table with only cells and genes given by filtering thresholds. A pdf showing the effect of genes counts of the filtering and a filteredStatistics.txt indicating how many cell and genes were filtered out</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Raffaele Calogero, Luca Alessandri</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>        <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget http://130.192.119.59/public/testSCumi_mm10.csv.zip"</span><span class="op">)</span></span></span>
<span class="r-in"><span>     <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rCASC</span><span class="op">)</span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"unzip testSCumi_mm10.csv.zip"</span><span class="op">)</span></span></span>
<span class="r-in"><span>     <span class="co">#filtering low quality cells</span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="lorenzFilter.html">lorenzFilter</a></span><span class="op">(</span>group<span class="op">=</span><span class="st">"docker"</span>,scratch.folder<span class="op">=</span><span class="st">"/data/scratch/"</span>,</span></span>
<span class="r-in"><span>                  file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"testSCumi_mm10.csv"</span>,sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  p_value<span class="op">=</span><span class="fl">0.05</span>,separator<span class="op">=</span><span class="st">','</span><span class="op">)</span></span></span>
<span class="r-in"><span>     <span class="co">#running annotation and removal of mit and ribo proteins genes</span></span></span>
<span class="r-in"><span>     <span class="co">#download mouse GTF for mm10</span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget ftp://ftp.ensembl.org/pub/release-92/gtf/mus_musculus/Mus_musculus.GRCm38.92.gtf.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"gunzip Mus_musculus.GRCm38.92.gtf.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span>     <span class="fu">scannobyGtf</span><span class="op">(</span>group<span class="op">=</span><span class="st">"docker"</span>, file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"testSCumi_mm10.csv"</span>,sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  gtf.name<span class="op">=</span><span class="st">"Mus_musculus.GRCm38.94.gtf"</span>, biotype<span class="op">=</span><span class="st">"protein_coding"</span>, </span></span>
<span class="r-in"><span>                  mt<span class="op">=</span><span class="cn">TRUE</span>, ribo.proteins<span class="op">=</span><span class="cn">TRUE</span>, umiXgene<span class="op">=</span><span class="fl">3</span>, riboStart.percentage<span class="op">=</span><span class="fl">0</span>, </span></span>
<span class="r-in"><span>                  riboEnd.percentage<span class="op">=</span><span class="fl">100</span>, mitoStart.percentage<span class="op">=</span><span class="fl">0</span>, mitoEnd.percentage<span class="op">=</span><span class="fl">100</span>, thresholdGenes<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Raffaele Calogero, Luca Alessandri, Marco Beccuti.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

